File: examples/canonical-terminal/elixir-optionparser-owl/lib/taskflow/core/filter.ex
defmodule Taskflow.Core.Filter do
  def filter(jobs, opts \\ []) do
    jobs
    |> maybe_filter_status(Keyword.get(opts, :status))
    |> maybe_filter_tag(Keyword.get(opts, :tag))
  end

  def sort(jobs), do: Enum.sort_by(jobs, &{&1.started_at || ~U[1970-01-01 00:00:00Z], &1.name}, :desc)

  defp maybe_filter_status(jobs, nil), do: jobs
  defp maybe_filter_status(jobs, ""), do: jobs
  defp maybe_filter_status(jobs, status), do: Enum.filter(jobs, &(&1.status == status))

  defp maybe_filter_tag(jobs, nil), do: jobs
  defp maybe_filter_tag(jobs, ""), do: jobs
  defp maybe_filter_tag(jobs, tag), do: Enum.filter(jobs, &(tag in &1.tags))
end

File: examples/canonical-terminal/ruby-clamp-tty/lib/taskflow/core/filter.rb
# frozen_string_literal: true

module Taskflow
  module Core
    module Filter
      module_function

      def filter(jobs, status: nil, tag: nil)
        filtered = jobs
        filtered = filtered.select { |job| job.status == status } if status && !status.empty?
        filtered = filtered.select { |job| job.tags.include?(tag) } if tag && !tag.empty?
        filtered
      end

      def sort(jobs)
        jobs.sort_by { |job| [job.started_at || Time.at(0), job.name] }.reverse
      end
    end
  end
end
